LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/JoinDeviceLoader.cs

```diff
@@ -3,56 +3,71 @@
 
 namespace LoRaWan.NetworkServer
 {
+    using System;
+    using System.Threading;
     using System.Threading.Tasks;
     using Microsoft.Extensions.Logging;
 
     /// <summary>
     /// Loads devices for join requests
     /// This object remains in cache until join succeeds or timeout expires (2 minutes).
     /// </summary>
-    internal class JoinDeviceLoader
+    internal class JoinDeviceLoader : IDisposable
     {
         private readonly IoTHubDeviceInfo ioTHubDevice;
         private readonly ILoRaDeviceFactory deviceFactory;
-        private readonly ILogger<JoinDeviceLoader> logger;
-        private readonly Task<LoRaDevice> loading;
+        private readonly LoRaDeviceCache deviceCache;
         private volatile bool canCache;
+        private SemaphoreSlim joinLock = new SemaphoreSlim(1);
+        private readonly ILogger<JoinDeviceLoader> logger;
 
         internal bool CanCache => this.canCache;
 
-        internal JoinDeviceLoader(IoTHubDeviceInfo ioTHubDevice, ILoRaDeviceFactory deviceFactory, ILogger<JoinDeviceLoader> logger)
+        internal JoinDeviceLoader(IoTHubDeviceInfo ioTHubDevice, ILoRaDeviceFactory deviceFactory, LoRaDeviceCache deviceCache, ILogger<JoinDeviceLoader> logger)
         {
             this.ioTHubDevice = ioTHubDevice;
             this.deviceFactory = deviceFactory;
+            this.deviceCache = deviceCache;
             this.logger = logger;
             this.canCache = true;
-            this.loading = Task.Run(() => LoadAsync());
         }
 
-        /// <summary>
-        /// Returns Task containing the device loading execution.
-        /// Waiting for it will suspend your thread/task until the device join is complete.
-        /// </summary>
-        internal Task<LoRaDevice> WaitCompleteAsync() => this.loading;
-
-        private async Task<LoRaDevice> LoadAsync()
+        internal async Task<LoRaDevice> LoadAsync()
         {
-            var loRaDevice = this.deviceFactory.Create(this.ioTHubDevice);
-
-            if (await loRaDevice.InitializeAsync())
+            await this.joinLock.WaitAsync();
+            try
             {
-                return loRaDevice;
+                if (this.deviceCache.TryGetByDevEui(this.ioTHubDevice.DevEUI, out var cachedDevice))
+                {
+                    return cachedDevice;
+                }
+                return await this.deviceFactory.CreateAndRegisterAsync(this.ioTHubDevice, CancellationToken.None);
```

Comment:
I have a question regarding join requests let's discuss

Reply:
We discussed, we will add a clear cache on deveui when a join request is received

Reply:
tracked in #1024